ramips: Adjust partition layout on Edgerouter-X to allow for kernels up to 6MB

darkxst opened the pull request:
The OEM layout on the Edgerouter X has two kernel slots that are 3MB each. Starting with Linux 6.1 the kernel images no longer fit into this layout. Currently the Edgerouter-X image builds are disabled due to this issue.

This PR will convert the Edgerouter X to use a single kernel slot that is 6MB, this is achieved with the following changes:
* Adjust dts to make `kernel1` 6MB and drop `kernel2` slot
* Patch `ubnt.sh` to always flash to `kernel1` slot (Likewise set the flag in factory partition to make sure kernel1 slot is active). This file is specific only to Edgerouter X and X-SFP. U-boot will now always boot from kernel1 slot, there is no failover logic within u-boot itself, that could change this boot setting.
* Set compat version to 2.0, to ensure the system has the updated `ubnt.sh` providing support for the new layout.

I have a second [PR](https://github.com/openwrt/openwrt/pull/15197)  that provides a migration script, that can hopefully be backported to 23.05, allowing users an easy path to migrate to the new layout.

I have also confirmed there is no issue downgrading back to 23.05 from this updated layout, where it will just go back to toggling kernel slots as per normal.

I have tested on Edgerouter X, however same changes also apply to Edgerouter X-SFP which will work the same.

OEM layout:
```
root@OpenWrt:/proc# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00080000 00020000 "u-boot"
mtd1: 00060000 00020000 "u-boot-env"
mtd2: 00060000 00020000 "factory"
mtd3: 00300000 00020000 "kernel1"
mtd4: 00300000 00020000 "kernel2"
mtd5: 0f7c0000 00020000 "ubi"
```

New layout:
```
root@OpenWrt:/# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00080000 00020000 "u-boot"
mtd1: 00060000 00020000 "u-boot-env"
mtd2: 00060000 00020000 "factory"
mtd3: 00600000 00020000 "kernel"
mtd4: 0f7c0000 00020000 "ubi"
```

darkxst committed:
rampis: mt7621: Convert Edgerouter-X dts to single kernel config

The OEM layout for Edgerouter X provides for two 3MB kernel slots.
As of linux 6.1 the kernel images no longer fit and as such
Edgerouter X builds have been disabled in Main.

Revise the layout to make kernel1 slot 6MB and drop kernel2 slot.

This patch applies the required changes to the dts file.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

darkxst committed:
rampis: mt7621: resize Edgerouter-X kernel size to 6MB

With the updated partition layout set in dts, set the KERNEL_SIZE
parameter to 6MB allowing builds of Linux 6.1 and later to fit.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

darkxst committed:
rampis: mt7621: Edgerouter-X disable toggling of kernel slots

Uboot selects which kernel slot to boot based on a flag in the factory
mtd partition. Patch ubnt.sh to ensure always flash to kernel1 slot and
update flag if required.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

darkxst committed:
rampis: mt7621: Re-enable Edgerouter-X image builds

With the new layout providing 6MB for kernel there will be no issues
with kernel size affecting build of images.

Re-enable image builds for Edgerouter-X and X-SFP.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

darkxst committed:
rampis: mt7621: Edgerouter-X set compat version message

Refer user to the wiki page for instructions on how to migrate to the
new kernel layout.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

darkxst committed:
rampis: mt7621: Edgerouter-X update compat version to 2.0

Use compat version to indicate that the new layout for larger kernels
is in place. This handled by the patch to ubnt.sh to always select
the kernel1 slot for flashing and as active kernel slot.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

DragonBluep commented on target/linux/ramips/dts/mt7621_ubnt_edgerouter-x.dtsi:
```diff
@@ -101,12 +101,7 @@
 
 		partition@140000 {
 			label = "kernel1";
```
If you want to merge these two partitions, why not use a more generic name `kernel`.

darkxst replied:
yes good point

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2506,12 +2506,16 @@ define Device/ubnt_edgerouter_common
   IMAGE_SIZE := 256768k
   FILESYSTEMS := squashfs
-  KERNEL_SIZE := 3145728
+  KERNEL_SIZE := 6291456
```
```suggestion
  KERNEL_SIZE := 6144k
```

DragonBluep commented on target/linux/ramips/mt7621/base-files/etc/uci-defaults/05_fix-compat-version:
```diff
@@ -0,0 +1,11 @@
+. /lib/functions.sh
+
+case "$(board_name)" in
```
This file seems unnecessary. The config in `05_compat-version` should be enough.

darkxst replied:
The config in `05_compat-version` only gets applied when using `sysupgrade -n` to discard settings. If user upgrades keeping settings without this, compat version will stay on 1.1 even though they have the new changes.

DragonBluep replied:
But that won't happen because sysupgrade will surely brick their devices due to small kernel partition. We expect a new installation here, right? (Though user can manually restore their settings, then compat_version will roll back to 1.1 again, too bad.) If I'm wrong please correct me.

And I feel attaching a migration firmware to this PR or the wiki would be more useful than modifying the v23.05 branch. In this way user can find a specific file to update their older version, v21.x, v22.x and v23.x. We can ask user search wiki in the `DEVICE_COMPAT_MESSAGE`.

darkxst replied:
> But that won't happen because sysupgrade will surely brick their devices due to small kernel partition.

Sysupgrade directly to a release with kernel more than 3MB will brick. That was the idea of having a transitional step through 23.x. Provided the new kernel is < 3MB (which it is in 23.x), then:

- First sysupgrade will update `ubnt.sh` and apply the new kernel layout from dts. Which is safe even if the kernel lands in kernel slot 2.
- The next subsequent update kernel will be installed to slot 1 (without the 3MB limitation.)
- User can keep settings through both these upgrades.


However I guess, as you mention, this could be done as a separate firmware build attached to wiki rather than changing 23.05. 

So i will drop this change from this PR.

DragonBluep commented on target/linux/ramips/mt7621/base-files/lib/upgrade/ubnt.sh:
```diff
@@ -8,32 +8,23 @@
-ubnt_get_target_kernel() {
+ubnt_get_kernel_flag() {
 	local factory_mtd=$1
 	local current_kernel_index=$(hexdump -s $UBNT_ERX_KERNEL_INDEX_OFFSET -n 1 -e '/1 "%X "' ${factory_mtd})
```
This tiny function can be merged into `ubnt_update_kernel_flag()`.

darkxst replied:
sure

DragonBluep reviewed (commented):
And if v23.05.x is still work, we'd better not to touch it.

darkxst commented:
> And if v23.05.x is still work, we'd better not to touch it.

The changes to v23.05.x would make upgrades to v24.x possible. Without that the only option to install new layout is to use serial console to tftp boot new initramfs image and sysupgrade from there.

darkxst committed:
ramips: mt7621: Edgerouter-X drop factory image

Edgerouter-X factory images have not built automatically since 19.x due
to images being over 3MB. While it was possible to build custom images
with very stripped down config, this is no longer possible with the size
increases of linux 6.1 and 6.6.

Drop code for generation of factory images, if some dev later wishes to
try custom images they can revert this commit.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

DragonBluep commented on target/linux/ramips/mt7621/base-files/lib/upgrade/ubnt.sh:
```diff
@@ -8,32 +8,16 @@
 	fi
 
-	if ! (echo -e $new_kernel_index | dd of=${factory_mtd} bs=1 count=1 seek=$UBNT_ERX_KERNEL_INDEX_OFFSET); then
+	if ! (echo -e "\x00" | dd of=${factory_mtd} bs=1 count=1 seek=$UBNT_ERX_KERNEL_INDEX_OFFSET); then
```
The original code seems have a BUG here as dd command can't flip Flash bit from ‘0’ to ‘1’. However, it should be fine now.

darkxst replied:
I pretty sure the kernel flag is a full byte and not a single bit.

DragonBluep replied:
I mean we must erase Flash block/byte before write operation. dd can only be used to write eMMC Flash. However, writing byte 0x00 is a special case.

darkxst replied:
ah ok

DragonBluep commented on target/linux/ramips/mt7621/base-files/lib/upgrade/ubnt.sh:
```diff
@@ -46,14 +30,8 @@ platform_upgrade_ubnt_erx() {
-
 	# This is a global defined in nand.sh, sets partition kernel will be flashed into
-	CI_KERNPART=${kernel_part}
+	CI_KERNPART="kernel"
```
I guess this line can be removed. I remember `kernel` is the default value.

darkxst commented:
Migration builds based on 23.05.3 with necessary patches to transition to the new layout proposed here, are located at:
https://github.com/darkxst/openwrt/releases/tag/migration-v0.1

msandber commented:
> And if v23.05.x is still work, we'd better not to touch it.

My two cents are, that in a sense v23.05.x for Edgerouter X is broken because it does not allow upgrading to the next major release. Resorting to an external image for conversion is clumsier and creates worse user experience.

dkadioglu commented:
If I may, I would like to ask, if the partition layout could also be modified as such that we could have an a/b layout to switch between two firmware images. Like described here https://docs.openwrt.melmac.net/luci-mod-alt-reboot/
I have no idea, what requirements a device, bootloader etc. have to fulfill to support this, so maybe it is not possible by design. Therefor, I would be thankful for a yes or no and some insight/background. Thanks.

darkxst commented:
> modified as such that we could have an a/b layout to switch between two firmware images. Like described here https://docs.openwrt.melmac.net/luci-mod-alt-reboot/

That mod is specifically for devices that already have a dual partition layout.  The Edgerouter-X only has a single shared rootfs, for the dual kernel slots, however now with these changes we are left with just the single kernel slot. Unfortunately not much else is possible with the vendor provided uboot.

dkadioglu commented:
> > modified as such that we could have an a/b layout to switch between two firmware images. Like described here https://docs.openwrt.melmac.net/luci-mod-alt-reboot/
> 
> That mod is specifically for devices that already have a dual partition layout. The Edgerouter-X only has a single shared rootfs, for the dual kernel slots, however now with these changes we are left with just the single kernel slot. Unfortunately not much else is possible with the vendor provided uboot.

Thank you very much for clarifying! Was just wondering, if that safety net feature would be possible, too.

darkxst commented:
If you just want a safety net, get yourself a USB-UART adaptor to connect to serial port and recover, if you somehow mess up openwrt!

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2506,12 +2482,13 @@ define Device/ubnt_edgerouter_common
   DEVICE_VENDOR := Ubiquiti
   IMAGE_SIZE := 256768k
```
`259840k`

darkxst replied:
I have updated this.

darkxst committed:
rampis: mt7621: Edgerouter-X adjust IMAGE_SIZE to reflect new kernel size

IMAGE_SIZE was previously set to kernel1 + ubi size = 256768k, now
kernel1 is 6MB adjust this value to add 3072k to total image size.

Signed-off-by: Tim Lunn <tim@feathertop.org>
Link: https://github.com/openwrt/openwrt/pull/15194
Signed-off-by: Petr Štetiar <ynezz@true.cz>

darkxst commented:
some test results from another thread, with this branch and the 23.05 backports in #15197 
> Tested:
OpenWrt 23.x -> main with 6MB slot. It gave a warning as expected.
OpenWrt 23.x -> 23.x with 6MB slot landing on 'kernel2'. Works as expected.
23x with 6Mb slot -> main with 6MB slot. Works as expected
main with 6MB slot -> OpenWrt 23.x. Forced, but was able to downgrade. Worked as expected.

> Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>

https://github.com/darkxst/openwrt/commit/ef1747cb77a6d76843d6a8f2d39e98a686d029f7#commitcomment-141174258

ynezz commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2504,14 +2480,15 @@ define Device/ubnt_edgerouter_common
-  DEFAULT := n
+  DEVICE_COMPAT_VERSION := 2.0
+  DEVICE_COMPAT_MESSAGE :=  Partition table has been changed due to kernel size restrictions. \
+    Refer to the wiki page for instructions to migrate to the new layout: \
```
What is the plan here? I couldn't find any instructions, but I might be blind :)

darkxst replied:
@ynezz I will add instructions on wiki page, as soon as its merged into main (or just prior). Didnt want to confuse people with documentation for the new layout when it wasn't available yet.

darkxst replied:
> What is the plan here? I couldn't find any instructions

I added some initial notes to the wiki. Of course until this PR is merged there are not actually any images available with the new kernel layout and hence also no actual images for the migration scripts to download...

https://openwrt.org/toh/ubiquiti/edgerouter_x_er-x_ka#migrating_to_linux_66_for_24xx

lantis1008 replied:
What is the expected user experience? Sysupgrade without preserve settings or sysupgrade a factory image?
It may be better to include this in the compat message and the wiki can be additional information. If the wiki became unavailable this is self contained. 
Other compat messages include this basic level of information and I think it is a good precedent to follow. 

My 2cents.

tl-sl replied:
A direct sysupgrade is not possible as it will truncate the kernel at 3MB and brick the router. Factory images are also very constrained on this device, it is no longer possible to build a working factory image since a very long time. Previously there have been seperate community build factory images with minimal configs, but even that wont be possible with 24.xx and Linux 6.6. 

This leaves two options to upgrade

1. Run migration script as proposed in #15197 for backport to 23.05.x, this takes care of flashing the new kernel across both existing kernel partitions and migrate system without preserving settings.
2. Reinstall a fresh install from serial console and TFTP to boot initramfs.

Documenting the process clearly at this stage is hard until I know exactly what is happening with both of these PR's. Hence the somewhat vague compat message. I will happily circle back around and clarify documentation in the compat message (and wiki), once i have a clear idea of what will or what might not be merged.

darkxst replied:
A direct sysupgrade is not possible as it will truncate the kernel at 3MB and brick the router. Factory images are also very constrained on this device, it is no longer possible to build a working factory image since a very long time. Previously there have been seperate community build factory images with minimal configs, but even that wont be possible with 24.xx and Linux 6.6.

This leaves two options to upgrade

* Run migration script as proposed in #15197  for backport to 23.05.x, this takes care of flashing the new kernel across both existing kernel partitions and migrate system without preserving settings.

* Reinstall a fresh install from serial console and TFTP to boot initramfs.

Documenting the process clearly at this stage is hard until I know exactly what is happening with both of these PR's. Hence the somewhat vague compat message. I will happily circle back around and clarify documentation in the compat message (and wiki), once i have a clear idea of what will or what might not be merged.

ynezz reviewed (approved):
Thanks, it seems there's no other or better way to maintain support for this device, so it looks good to me.

darkxst commented:
> Thanks, it seems there's no other or better way to maintain support for this device, so it looks good to me.

I will point out that this device never had a proper dual firmware layout, it was only dual kernel, with a single shared rootfs. Further there was no mechanism to fail over to the other kernel (other than via serial console uboot shell). So probably not loosing much here.

ptpt52 commented:
ping @981213 
is it possible to use a second u-boot for this device? just like a6991fc7d251f7ab65a588546e22abfd4f8ce472

darkxst commented:
> is it possible to use a second u-boot for this device? 

In theory I suppose it should be possible to add a second u-boot, however IMO it adds a lot of complexity to both the build and the migration on upgrade, for very little benefit. 6MB should be plenty of space for the kernel for quite some time going forward.

darkxst commented:
My other PR now includes a standalone migration script rather than patching sysupgrade on stable releases.

https://github.com/openwrt/openwrt/pull/15197

msandber commented:
> ping @981213 is it possible to use a second u-boot for this device? just like [a6991fc](https://github.com/openwrt/openwrt/commit/a6991fc7d251f7ab65a588546e22abfd4f8ce472)

I kinda like this idea. The second u-boot is likely to get updated less often than the kernel, if ever. So wear-leveling-wise it makes sense to place kernel and uboot env on ubi.

On the other hand, the JTAG interface looks to be operational so replacing the u-boot should be possible too. U-boot supports mt7621, there is an example for one such device under boot-mediatek and there exists openocd configs [here](https://github.com/mtk-openwrt/openocd-scripts).

I am afraid my JTAG cable/adapter (bus blaster 3) might not be up to the job. It finds 4-5 chains but that's it I haven't figured out how to backup/restore original u-boot should I mess it up. 

@hackpascal would you have pointers on which JTAG dongle works and how to configure nand flash or write it some other way?

hackpascal commented:
> It finds 4-5 chains

MT7621 does have 5 TAPs. Your bus blaster v3 works as expected.

> @hackpascal would you have pointers on which JTAG dongle works and how to configure OpenOCD for nand flash or write it some other way?

Please refer to https://github.com/mtk-openwrt/openocd-scripts/tree/main/mt7621
The script is tested with bus blaster v4 😃

ptpt52 commented:
I think we may not need to switch to new layout if build the firmware with `CONFIG_KERNEL_KALLSYMS` disabled, this would reduce the kernel size to 2.8M

darkxst commented:
> I think we may not need to switch to new layout if build the firmware with `CONFIG_KERNEL_KALLSYMS` disabled, this would reduce the kernel size to 2.8M

This seems like a somewhat temporary solution until presumably the kernel again goes over 3MB again a later release cycle.

The dual kernel slots are not really used anyway, there is no failover to the other kernel if boot fails. The only way to change kernel slot is logging into the u-boot serial console and manually flipping the flag. Also there is no redundant rootfs either.

carlo2007 commented:
The pull request is not merged for the open issue [15643](https://github.com/openwrt/openwrt/issues/15643).
It seems that the issue was introduced by commit [2a25c6a](https://github.com/openwrt/openwrt/commit/2a25c6ace8d833cf491a66846a0b9e7c5387b8f0). I reverted the commit [2a25c6a](https://github.com/openwrt/openwrt/commit/2a25c6ace8d833cf491a66846a0b9e7c5387b8f0) and applied the patch set of this pull request and everything worked correctly.

@darkxst please consider to add the revert of the commit [2a25c6a](https://github.com/openwrt/openwrt/commit/2a25c6ace8d833cf491a66846a0b9e7c5387b8f0) to this pull request. In this way the pull request will pass all the checks.

LynxChaus commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2562,14 +2538,15 @@ define Device/ubnt_edgerouter_common
   FILESYSTEMS := squashfs
-  KERNEL_SIZE := 3145728
-  KERNEL_INITRAMFS := $$(KERNEL) | \
-	ubnt-erx-factory-image $(KDIR)/tmp/$$(KERNEL_INITRAMFS_PREFIX)-factory.tar
```
After removing specific image build target - size of image not checked at all.

remove KERNEL_SIZE define, set IMAGE_SIZE to 6144k, and modify ``IMAGE/sysupgrade.bin`` target to 
``IMAGE/sysupgrade.bin := sysupgrade-tar | check-size | append-metadata``

darkxst replied:
Is the sysupgrade.bin limited to 6MB though?  I would have thought it isn't.

LynxChaus replied:
kernel + squashfs image should fit in ``kernel`` partition. Or we need use mtdsplit partition scheme for floating ``ubi`` partition.

darkxst replied:
factory flashing is not possible anymore as that is limited to 3MB still and no image can be built that small anymore. It is required to flash older version first or use tftp to install.

LynxChaus replied:
Well, I'm wrong about _kernel + squashfs image should fit in kernel partition_, but simply set ``IMAGE_SIZE`` define - do nothing. At least - you need manually check via ``check-size`` size of root squashfs.

PS: And 259840+6144 is greater than usable RAM on device. and twice greater than /tmp partition.

darkxst commented:
> or will this PR will make it into the just branched 24.10 release?

I hope it can still be merged. However I have no idea if it actually will, its been a very long time already since I created this PR.

systemcrash commented:
This PR makes sense, and should extend the support life of this device a while longer.


I don't yet know enough to advise on the topic, but perhaps a UBI layout would also work? UBI likely requires support from uboot. It would appear that a bunch of ramips mt7621 devices support UBI, so it's not unreasonable to think that this one can.

darkxst commented:
> Edit: since your partition layout already includes ubi, why not, provided the uboot can boot the kernel from a UBI.

Ubiquity u-boot does not have any support for UBI, and in fact their entire solution to A/B boot and recovery images is completely non-standard homemade cruft. They also refused to release any GPL source code to me when I requested it.

It would certainly be possible to build a custom u-boot, however this would significantly complicate the migration process to Openwrt and/or back to factory firmware.

LynxChaus commented:
mainline u-boot already have support for mt7621 boards (from 2022) with/without NAND, limited support for ar803x and pca953x.
